File: lib/presentation/login_screen/login_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:sizer/sizer.dart';

import '../../core/app_export.dart';
import '../../theme/app_theme.dart';
import '../../widgets/custom_icon_widget.dart';
import './widgets/battery_loading_widget.dart';
import './widgets/lock_in_logo_widget.dart';
import './widgets/login_form_widget.dart';
import './widgets/social_login_widget.dart';
import 'widgets/battery_loading_widget.dart';
import 'widgets/lock_in_logo_widget.dart';
import 'widgets/login_form_widget.dart';
import 'widgets/social_login_widget.dart';

class LoginScreen extends StatefulWidget {
  const LoginScreen({super.key});

  @override
  State<LoginScreen> createState() => _LoginScreenState();
}

class _LoginScreenState extends State<LoginScreen> {
  bool _isLoading = false;
  double _loadingProgress = 0.0;
  String _errorMessage = '';

  // Mock credentials for testing
  final Map<String, String> _mockCredentials = {
    'admin@lockin.com': 'admin123',
    'user@lockin.com': 'user123',
    'demo@lockin.com': 'demo123',
  };

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: AppTheme.lightTheme.scaffoldBackgroundColor,
      body: SafeArea(
        child: GestureDetector(
          onTap: () => FocusScope.of(context).unfocus(),
          child: SingleChildScrollView(
            physics: ClampingScrollPhysics(),
            child: ConstrainedBox(
              constraints: BoxConstraints(
                minHeight: MediaQuery.of(context).size.height -
                    MediaQuery.of(context).padding.top -
                    MediaQuery.of(context).padding.bottom,
              ),
              child: Padding(
                padding: EdgeInsets.symmetric(horizontal: 6.w),
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    SizedBox(height: 8.h),

                    // Lock In Logo with gamified styling
                    LockInLogoWidget(),

                    SizedBox(height: 6.h),

                    // Welcome text
                    Text(
                      'Welcome Back',
                      style: AppTheme.lightTheme.textTheme.headlineMedium
                          ?.copyWith(
                        fontWeight: FontWeight.w700,
                        color: AppTheme.lightTheme.colorScheme.onSurface,
                      ),
                    ),

                    SizedBox(height: 1.h),

                    Text(
                      'Continue your skill development journey',
                      style: AppTheme.lightTheme.textTheme.bodyLarge?.copyWith(
                        color: AppTheme.lightTheme.colorScheme.onSurfaceVariant,
                      ),
                      textAlign: TextAlign.center,
                    ),

                    SizedBox(height: 4.h),

                    // Battery Loading Widget
                    BatteryLoadingWidget(
                      isVisible: _isLoading,
                      progress: _loadingProgress,
                    ),

                    if (_isLoading) SizedBox(height: 3.h),

                    // Error Message
                    if (_errorMessage.isNotEmpty) ...[
                      Container(
                        width: double.infinity,
                        padding: EdgeInsets.all(16),
                        margin: EdgeInsets.only(bottom: 3.h),
                        decoration: BoxDecoration(
                          color: AppTheme.warningLight.withValues(alpha: 0.1),
                          borderRadius: BorderRadius.circular(12),
                          border: Border.all(
                            color: AppTheme.warningLight.withValues(alpha: 0.3),
                            width: 1,
                          ),
                        ),
                        child: Row(
                          children: [
                            CustomIconWidget(
                              iconName: 'error_outline',
                              color: AppTheme.warningLight,
                              size: 20,
                            ),
                            SizedBox(width: 12),
                            Expanded(
                              child: Text(
                                _errorMessage,
                                style: AppTheme.lightTheme.textTheme.bodyMedium
                                    ?.copyWith(
                                  color: AppTheme.warningLight,
                                ),
                              ),
                            ),
                          ],
                        ),
                      ),
                    ],

                    // Login Form
                    LoginFormWidget(
                      onLogin: _handleLogin,
                      isLoading: _isLoading,
                    ),

                    SizedBox(height: 4.h),

                    // Social Login
                    SocialLoginWidget(
                      onSocialLogin: _handleSocialLogin,
                      isLoading: _isLoading,
                    ),

                    SizedBox(height: 4.h),

                    // Sign up link
                    Row(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        Text(
                          'New to Lock In? ',
                          style: AppTheme.lightTheme.textTheme.bodyMedium
                              ?.copyWith(
                            color: AppTheme
                                .lightTheme.colorScheme.onSurfaceVariant,
                          ),
                        ),
                        GestureDetector(
                          onTap: _isLoading
                              ? null
                              : () {
                                  Navigator.pushNamed(
                                      context, '/register-screen');
                                },
                          child: Text(
                            'Start Journey',
                            style: AppTheme.lightTheme.textTheme.bodyMedium
                                ?.copyWith(
                              color: AppTheme.lightTheme.colorScheme.primary,
                              fontWeight: FontWeight.w600,
                            ),
                          ),
                        ),
                      ],
                    ),

                    SizedBox(height: 4.h),
                  ],
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }

  void _handleLogin(String email, String password) async {
    setState(() {
      _isLoading = true;
      _errorMessage = '';
      _loadingProgress = 0.0;
    });

    // Simulate authentication process with battery loading
    for (int i = 0; i <= 100; i += 10) {
      await Future.delayed(Duration(milliseconds: 80));
      if (mounted) {
        setState(() {
          _loadingProgress = i / 100.0;
        });
      }
    }

    // Check mock credentials
    if (_mockCredentials.containsKey(email.toLowerCase()) &&
        _mockCredentials[email.toLowerCase()] == password) {
      // Success
      HapticFeedback.heavyImpact();

      if (mounted) {
        setState(() {
          _isLoading = false;
        });

        // Navigate to dashboard
        Navigator.pushReplacementNamed(context, '/dashboard-home');
      }
    } else {
      // Error
      HapticFeedback.vibrate();

      if (mounted) {
        setState(() {
          _isLoading = false;
          _loadingProgress = 0.0;
          _errorMessage = _getErrorMessage(email, password);
        });
      }
    }
  }

  String _getErrorMessage(String email, String password) {
    if (!_mockCredentials.containsKey(email.toLowerCase())) {
      return 'Account not found. Please check your email or create a new account.';
    } else {
      return 'Incorrect password. Please try again or reset your password.';
    }
  }

  void _handleSocialLogin(String provider) async {
    setState(() {
      _isLoading = true;
      _errorMessage = '';
      _loadingProgress = 0.0;
    });

    // Simulate social login process
    for (int i = 0; i <= 100; i += 20) {
      await Future.delayed(Duration(milliseconds: 100));
      if (mounted) {
        setState(() {
          _loadingProgress = i / 100.0;
        });
      }
    }

    // Simulate successful social login
    HapticFeedback.heavyImpact();

    if (mounted) {
      setState(() {
        _isLoading = false;
      });

      // Navigate to dashboard
      Navigator.pushReplacementNamed(context, '/dashboard-home');
    }
  }
}

File: src/screens/LoginScreen.js
import React, { useState } from 'react';
import {
  View,
  Text,
  TextInput,
  TouchableOpacity,
  StyleSheet,
  ScrollView,
  Alert,
  KeyboardAvoidingView,
  Platform,
  useColorScheme,
} from 'react-native';
import { useNavigation } from '@react-navigation/native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import HapticFeedback from 'react-native-haptic-feedback';

import CustomIcon from '../components/CustomIcon';
import BatteryProgressIndicator from '../components/BatteryProgressIndicator';
import { Colors, Typography, Spacing, BorderRadius, responsiveWidth, responsiveHeight, DarkTheme } from '../theme/AppTheme';

const LoginScreen = () => {
  const navigation = useNavigation();
  const [isLoading, setIsLoading] = useState(false);
  const [loadingProgress, setLoadingProgress] = useState(0);
  const [errorMessage, setErrorMessage] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const isDarkMode = useColorScheme() === 'dark';
  const theme = isDarkMode ? DarkTheme : { colors: Colors };

  const mockCredentials = {
    'admin@lockin.com': 'admin123',
    'user@lockin.com': 'user123',
    'demo@lockin.com': 'demo123',
  };

  const handleLogin = async () => {
    if (!email || !password) {
      setErrorMessage('Please fill in all fields');
      return;
    }

    setIsLoading(true);
    setErrorMessage('');
    setLoadingProgress(0);

    // Simulate authentication process with battery loading
    for (let i = 0; i <= 100; i += 10) {
      await new Promise(resolve => setTimeout(resolve, 80));
      setLoadingProgress(i / 100);
    }

    // Check mock credentials
    if (mockCredentials[email.toLowerCase()] === password) {
      HapticFeedback.trigger('impactHeavy');
      
      await AsyncStorage.setItem('is_logged_in', 'true');
      await AsyncStorage.setItem('user_email', email);
      
      setIsLoading(false);
      navigation.replace('MainTabs');
    } else {
      HapticFeedback.trigger('impactMedium');
      
      setIsLoading(false);
      setLoadingProgress(0);
      setErrorMessage(getErrorMessage(email, password));
    }
  };

  const getErrorMessage = (email, password) => {
    if (!mockCredentials[email.toLowerCase()]) {
      return 'Account not found. Please check your email or create a new account.';
    } else {
      return 'Incorrect password. Please try again or reset your password.';
    }
  };

  const handleSocialLogin = async (provider) => {
    setIsLoading(true);
    setErrorMessage('');
    setLoadingProgress(0);

    // Simulate social login process
    for (let i = 0; i <= 100; i += 20) {
      await new Promise(resolve => setTimeout(resolve, 100));
      setLoadingProgress(i / 100);
    }

    HapticFeedback.trigger('impactHeavy');
    
    await AsyncStorage.setItem('is_logged_in', 'true');
    await AsyncStorage.setItem('user_email', `user@${provider}.com`);
    
    setIsLoading(false);
    navigation.replace('MainTabs');
  };

  return (
    <KeyboardAvoidingView 
      style={[styles.container, { backgroundColor: theme.colors.background }]} 
      behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
    >
      <ScrollView 
        contentContainerStyle={styles.scrollContent}
        keyboardShouldPersistTaps="handled"
      >
        <View style={styles.content}>
          {/* Lock In Logo */}
          <View style={styles.logoContainer}>
            <CustomIcon name="lock" size={responsiveWidth(20)} color={theme.colors.primary} />
            <Text style={[styles.logoText, Typography.headlineMedium, { color: theme.colors.primary }]}>
              Lock In
            </Text>
          </View>

          {/* Welcome text */}
          <Text style={[styles.welcomeText, Typography.headlineMedium, { color: theme.colors.text }]}>
            Welcome Back
          </Text>

          <Text style={[styles.subtitle, Typography.bodyLarge, { color: theme.colors.textSecondary }]}>
            Continue your skill development journey
          </Text>

          {/* Battery Loading Widget */}
          {isLoading && (
            <View style={styles.loadingContainer}>
              <BatteryProgressIndicator
                progress={loadingProgress}
                width={responsiveWidth(30)}
                height={responsiveHeight(8)}
                batteryColor={theme.colors.primary}
              />
            </View>
          )}

          {/* Error Message */}
          {errorMessage ? (
            <View style={[styles.errorContainer, { 
              backgroundColor: theme.colors.warning + '1A',
              borderColor: theme.colors.warning + '4D'
            }]}>
              <CustomIcon name="error-outline" size={20} color={theme.colors.warning} />
              <Text style={[styles.errorText, Typography.bodyMedium, { color: theme.colors.warning }]}>
                {errorMessage}
              </Text>
            </View>
          ) : null}

          {/* Login Form */}
          <View style={styles.formContainer}>
            <View style={styles.inputContainer}>
              <Text style={[styles.inputLabel, Typography.labelMedium, { color: theme.colors.textSecondary }]}>
                Email
              </Text>
              <TextInput
                style={[styles.input, Typography.bodyLarge, { 
                  backgroundColor: theme.colors.surface,
                  borderColor: theme.colors.border,
                  color: theme.colors.text
                }]}
                placeholder="Enter your email"
                placeholderTextColor={theme.colors.textSecondary}
                value={email}
                onChangeText={setEmail}
                keyboardType="email-address"
                autoCapitalize="none"
                autoCorrect={false}
                editable={!isLoading}
              />
            </View>

            <View style={styles.inputContainer}>
              <Text style={[styles.inputLabel, Typography.labelMedium, { color: theme.colors.textSecondary }]}>
                Password
              </Text>
              <TextInput
                style={[styles.input, Typography.bodyLarge, { 
                  backgroundColor: theme.colors.surface,
                  borderColor: theme.colors.border,
                  color: theme.colors.text
                }]}
                placeholder="Enter your password"
                placeholderTextColor={theme.colors.textSecondary}
                value={password}
                onChangeText={setPassword}
                secureTextEntry
                editable={!isLoading}
              />
            </View>

            <TouchableOpacity
              style={[styles.loginButton, { 
                backgroundColor: theme.colors.primary,
                opacity: isLoading ? 0.6 : 1 
              }]}
              onPress={handleLogin}
              disabled={isLoading}
            >
              <Text style={[styles.loginButtonText, Typography.labelLarge, { color: 'white' }]}>
                {isLoading ? 'Signing In...' : 'Sign In'}
              </Text>
            </TouchableOpacity>
          </View>

          {/* Social Login */}
          <View style={styles.socialContainer}>
            <Text style={[styles.socialText, Typography.bodyMedium, { color: theme.colors.textSecondary }]}>
              Or continue with
            </Text>

            <View style={styles.socialButtons}>
              <TouchableOpacity
                style={[styles.socialButton, { 
                  backgroundColor: theme.colors.surface,
                  borderColor: theme.colors.border,
                  opacity: isLoading ? 0.6 : 1 
                }]}
                onPress={() => handleSocialLogin('google')}
                disabled={isLoading}
              >
                <CustomIcon name="google" size={24} color={theme.colors.primary} />
                <Text style={[styles.socialButtonText, Typography.labelMedium, { color: theme.colors.primary }]}>
                  Google
                </Text>
              </TouchableOpacity>

              <TouchableOpacity
                style={[styles.socialButton, { 
                  backgroundColor: theme.colors.surface,
                  borderColor: theme.colors.border,
                  opacity: isLoading ? 0.6 : 1 
                }]}
                onPress={() => handleSocialLogin('apple')}
                disabled={isLoading}
              >
                <CustomIcon name="apple" size={24} color={theme.colors.text} />
                <Text style={[styles.socialButtonText, Typography.labelMedium, { color: theme.colors.text }]}>
                  Apple
                </Text>
              </TouchableOpacity>
            </View>
          </View>

          {/* Sign up link */}
          <View style={styles.signupContainer}>
            <Text style={[Typography.bodyMedium, { color: theme.colors.textSecondary }]}>
              New to Lock In?{' '}
            </Text>
            <TouchableOpacity
              onPress={() => navigation.navigate('Register')}
              disabled={isLoading}
            >
              <Text style={[Typography.bodyMedium, { color: theme.colors.primary, fontWeight: '600' }]}>
                Start Journey
              </Text>
            </TouchableOpacity>
          </View>
        </View>
      </ScrollView>
    </KeyboardAvoidingView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  scrollContent: {
    flexGrow: 1,
    justifyContent: 'center',
  },
  content: {
    paddingHorizontal: responsiveWidth(6),
    paddingVertical: responsiveHeight(4),
  },
  logoContainer: {
    alignItems: 'center',
    marginBottom: responsiveHeight(6),
  },
  logoText: {
    marginTop: Spacing.md,
    fontWeight: '700',
  },
  welcomeText: {
    fontWeight: '700',
    textAlign: 'center',
    marginBottom: responsiveHeight(1),
  },
  subtitle: {
    textAlign: 'center',
    marginBottom: responsiveHeight(4),
  },
  loadingContainer: {
    alignItems: 'center',
    marginBottom: responsiveHeight(3),
  },
  errorContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: Spacing.lg,
    borderRadius: BorderRadius.lg,
    borderWidth: 1,
    marginBottom: responsiveHeight(3),
  },
  errorText: {
    marginLeft: Spacing.md,
    flex: 1,
  },
  formContainer: {
    marginBottom: responsiveHeight(4),
  },
  inputContainer: {
    marginBottom: Spacing.lg,
  },
  inputLabel: {
    marginBottom: Spacing.sm,
  },
  input: {
    borderWidth: 1,
    borderRadius: BorderRadius.lg,
    paddingHorizontal: Spacing.lg,
    paddingVertical: Spacing.lg,
  },
  loginButton: {
    paddingVertical: Spacing.lg,
    borderRadius: BorderRadius.lg,
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: Spacing.md,
  },
  loginButtonText: {
    fontWeight: '600',
  },
  socialContainer: {
    alignItems: 'center',
    marginBottom: responsiveHeight(4),
  },
  socialText: {
    marginBottom: Spacing.lg,
  },
  socialButtons: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    width: '100%',
  },
  socialButton: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingHorizontal: Spacing.xl,
    paddingVertical: Spacing.lg,
    borderRadius: BorderRadius.lg,
    borderWidth: 1,
    minWidth: responsiveWidth(35),
    justifyContent: 'center',
  },
  socialButtonText: {
    marginLeft: Spacing.sm,
    fontWeight: '500',
  },
  signupContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export default LoginScreen;
